helix-core/src/test/java/org/apache/helix/controller/rebalancer/waged/model/TestClusterModel.java

```diff
@@ -0,0 +1,108 @@
+    Set<AssignableNode> nodeSet = new HashSet<>();
+    testCache.getInstanceConfigMap().values().stream().forEach(config -> nodeSet
+        .add(new AssignableNode(testCache, config.getInstanceName(), Collections.emptyList())));
+    return nodeSet;
+  }
+
+  @Test
+  public void testNormalUsage() throws IOException {
+    ResourceControllerDataProvider testCache = setupClusterDataCache();
+    Set<AssignableReplica> assignableReplicas = generateReplicas(testCache);
+    Set<AssignableNode> assignableNodes = generateNodes(testCache);
+
+    // Test 1 - initialization
+    ClusterContext context = new ClusterContext(assignableReplicas, 2);
+    ClusterModel clusterModel =
+        new ClusterModel(context, assignableReplicas, assignableNodes, Collections.emptyMap(),
+            Collections.emptyMap());
+
+    Assert.assertTrue(clusterModel.getContext().getAssignmentForFaultZoneMap().values().stream()
+        .allMatch(v -> v.values().isEmpty()));
+    Assert.assertFalse(clusterModel.getAssignableNodes().values().stream()
+        .anyMatch(n -> n.getCurrentAssignmentCount() != 0));
+
+    // The initialization of the context, node and replication has been tested separately. So for
+    // cluster model, focus on testing the assignment and release.
+
+    // Assign
+    AssignableReplica replica = assignableReplicas.iterator().next();
+    AssignableNode node = assignableNodes.iterator().next();
+    clusterModel
+        .assign(replica.getResourceName(), replica.getPartitionName(), replica.getReplicaState(),
+            node.getInstanceName());
+
+    Assert.assertTrue(
+        clusterModel.getContext().getAssignmentForFaultZoneMap().get(node.getFaultZone())
+            .get(replica.getResourceName()).contains(replica.getPartitionName()));
+    Assert.assertTrue(node.getCurrentAssignmentsMap().get(replica.getResourceName())
+        .contains(replica.getPartitionName()));
+
+    // Assign a non-exist replication
+    try {
+      clusterModel.assign("NOT-EXIST", replica.getPartitionName(), replica.getReplicaState(),
+          node.getInstanceName());
+      Assert.fail("Assigning a non existing resource partition shall fail.");
+    } catch (HelixException ex) {
+      // expected
+    }
+
+    // Assign a non-exist replication
+    try {
+      clusterModel
+          .assign(replica.getResourceName(), replica.getPartitionName(), replica.getReplicaState(),
+              "NON-EXIST");
+      Assert.fail("Assigning a resource partition to a non existing instance shall fail.");
+    } catch (HelixException ex) {
+      // expected
+    }
+
+    // Release
+    clusterModel
+        .release(replica.getResourceName(), replica.getPartitionName(), replica.getReplicaState(),
+            node.getInstanceName());
+
+    Assert.assertTrue(clusterModel.getContext().getAssignmentForFaultZoneMap().values().stream()
+        .allMatch(v -> v.values().stream().allMatch(s -> s.isEmpty())));
```

Comment:
Spell out one letter variables. They don't have to be long, but this is now affecting readability. Especially it becomes confusing when nested maps are involved.

Lambda expressions exist partly to reduce verbosity of Java but we shouldn't sacrifice readability.